src/crimson/os/seastore/seastore.cc

```diff
@@ -1041,141 +1040,191 @@ SeaStore::tm_ret SeaStore::_do_transaction_step(
+        return _setattrs(ctx, onodes[op->oid], std::move(to_set));
+      }
+      break;
+      case Transaction::OP_SETATTRS:
+      {
+        std::map<std::string, bufferlist> to_set;
+        i.decode_attrset(to_set);
+        return _setattrs(ctx, onodes[op->oid], std::move(to_set));
+      }
+      break;
+      case Transaction::OP_RMATTR:
+      {
+        std::string name = i.decode_string();
+        return _rmattr(ctx, onodes[op->oid], name);
+      }
+      break;
+      case Transaction::OP_RMATTRS:
+      {
+        return _rmattrs(ctx, onodes[op->oid]);
+      }
+      break;
+      case Transaction::OP_OMAP_SETKEYS:
+      {
+        std::map<std::string, ceph::bufferlist> aset;
+        i.decode_attrset(aset);
+        return _omap_set_values(ctx, onodes[op->oid], std::move(aset));
+      }
+      break;
+      case Transaction::OP_OMAP_SETHEADER:
+      {
+        ceph::bufferlist bl;
+        i.decode_bl(bl);
+        return _omap_set_header(ctx, onodes[op->oid], std::move(bl));
+      }
+      break;
+      case Transaction::OP_OMAP_RMKEYS:
+      {
+        omap_keys_t keys;
+        i.decode_keyset(keys);
+        return _omap_rmkeys(ctx, onodes[op->oid], std::move(keys));
+      }
+      break;
+      case Transaction::OP_OMAP_RMKEYRANGE:
+      {
+        string first, last;
+        first = i.decode_string();
+        last = i.decode_string();
+        return _omap_rmkeyrange(
+	  ctx, onodes[op->oid],
+	  std::move(first), std::move(last));
+      }
+      break;
+      case Transaction::OP_OMAP_CLEAR:
+      {
+        return _omap_clear(ctx, onodes[op->oid]);
+      }
+      break;
+      case Transaction::OP_ZERO:
+      {
+        objaddr_t off = op->off;
+        extent_len_t len = op->len;
+        return _zero(ctx, onodes[op->oid], off, len);
+      }
+      break;
+      default:
+        ERROR("bad op {}", static_cast<unsigned>(op->op));
+        return crimson::ct_error::input_output_error::make();
+      }
+    } catch (std::exception &e) {
+      ERROR("got exception {}", e);
       return crimson::ct_error::input_output_error::make();
     }
-  } catch (std::exception &e) {
-    ERROR("got exception {}", e);
-    return crimson::ct_error::input_output_error::make();
-  }
+  }).handle_error_interruptible(
+    tm_iertr::pass_further{},
+    crimson::ct_error::enoent::handle([op] {
+      if (op->op == Transaction::OP_CLONERANGE ||
+          op->op == Transaction::OP_CLONE ||
+          op->op == Transaction::OP_CLONERANGE2 ||
+          op->op == Transaction::OP_COLL_ADD ||
+          op->op == Transaction::OP_SETATTR ||
+          op->op == Transaction::OP_SETATTRS ||
+          op->op == Transaction::OP_RMATTR ||
+          op->op == Transaction::OP_OMAP_SETKEYS ||
+          op->op == Transaction::OP_OMAP_RMKEYS ||
+          op->op == Transaction::OP_OMAP_RMKEYRANGE ||
+          op->op == Transaction::OP_OMAP_SETHEADER) {
+        ceph_abort_msg("unexpected enoent error");
+      }
```

Comment:
Which op is allowed to throw the enoent error?

Seems to me in all cases, enoent is not expected.

Reply:
this follows bluestore and filestore's logic,  some ops such as  OMAP_CLEAR, TRUNCATE, REMOVE, will tolerate get_onode enoent.

Reply:
let's add comment here to clarify that OMAP_CLEAR, TRUNCATE, REMOVE will tolerate absent onode.